### PR TITLE
Add new check 'comparison-with-callable'

### PR DESCRIPTION
**Pull request by sushobhit27**
This check warn for comparison with bare callable, without calling it.

### Fixes / new features
- #2082

### CODE REVIEW

**Comment by sushobhit27** [comment]
@PCManticore Although I have taken care of all scenario in test cases, but I am not very sure about my way of getting all callable in module under this check.

**Comment by sushobhit27** [comment]
@PCManticore please review PR with latest changes.

File: pylint/checkers/base.py [hunk 1]
```diff
@@ -1810,9 +1816,35 @@ def _check_logical_tautology(self, node):
...
+
+        def _is_bare_callable(node):
+            bare_callables = (astroid.FunctionDef, astroid.BoundMethod)
+            if (isinstance(node, bare_callables) or
+                    isinstance(utils.safe_infer(node), bare_callables)):
+                return True
+            return False
+
+        left_operand, right_operand = node.left, node.ops[0][1]
+        bare_callables_count = [True for operand in (left_operand, right_operand)
```
Comment (PCManticore) [inline pylint/checkers/base.py]: Instead of holding a list, you can do something like `sum(True for operand in ...)` and then verify that the value is 1.

File: doc/whatsnew/2.0.rst [hunk 2]
```diff
@@ -25,6 +25,24 @@ Summary -- Release highlights
 
 New checkers
 ============
+* A new check was added, ``comparison-with-callable``.
+
+  This refactoring message is emitted when a comparison involves callable without parenthesis, which if compared
```
Comment (PCManticore) [inline doc/whatsnew/2.0.rst]: I'd rewrite this as:
`This refactoring message is emitted when pylint detects that a comparison with a callable was made, which might suggest that some paranthesis were omitted, resulting in potential unwanted behaviour`.

File: doc/whatsnew/2.0.rst [hunk 3]
```diff
@@ -25,6 +25,24 @@ Summary -- Release highlights
...
+  can result in unwanted behaviour.
+
+  .. code-block:: python
+    def foo():
+    return None
```
Comment (PCManticore) [inline doc/whatsnew/2.0.rst]: Need to indent this further.

File: pylint/checkers/base.py [hunk 4]
```diff
@@ -1731,6 +1731,12 @@ class ComparisonChecker(_BasicChecker):
                       'comparison-with-itself',
                       'Used when something is compared against itself.',
                       ),
+            'W0143': ('Comparing bare callable, might have skipped parenthesis after callable',
+                      'comparison-with-callable',
+                      'Used when callable is used in a comparison without parenthesis, '
```
Comment (PCManticore) [inline pylint/checkers/base.py]: I'd use the same message that I suggested in the release changelog.

File: pylint/checkers/base.py [hunk 5]
```diff
@@ -1731,6 +1731,12 @@ class ComparisonChecker(_BasicChecker):
                       'comparison-with-itself',
                       'Used when something is compared against itself.',
                       ),
+            'W0143': ('Comparing bare callable, might have skipped parenthesis after callable',
```
Comment (PCManticore) [inline pylint/checkers/base.py]: I'd just rephrase this as `Comparing against a callable, did you omit the parens?` I'm not entirely sold on the part after the comma, since we don't usually have messages that *talk* with the user like that, but can't think of something more suggestive.

Comment (sushobhit27) [reply pylint/checkers/base.py]: @PCManticore all points noted down. I will raise PR again in a day or two.
For msg, how about "Comparing against a callable, check parenthesis after callable"
Although did you omit... also seems correct to me as both messages points towards a probable root cause of the issue.

**Comment by sushobhit27** [comment]
@PCManticore it seems that CI tests are again failing due to some unrelated stuff. Could you pls check it.

**Comment by PCManticore** [comment]
@sushobhit27 I checked the PR, it looks great! Regarding the CI, the failure seems to be caused by this PR, except the linting job, which is broken at the moment. So we need to fix this before getting the patch in:
```
E           Failed: Wrong results for file "member_checks_opaque":
E           
E           Expected in testdata:
E              6: no-member
```

**Comment by PCManticore** [comment]
@sushobhit27 Actually it seems fine, as it is failing on master as well.
